spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/secretmanager/SecretManagerPropertySource.java

```diff
@@ -0,0 +1,110 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.autoconfigure.secretmanager;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.google.cloud.secretmanager.v1beta1.AccessSecretVersionResponse;
+import com.google.cloud.secretmanager.v1beta1.ProjectName;
+import com.google.cloud.secretmanager.v1beta1.Secret;
+import com.google.cloud.secretmanager.v1beta1.SecretManagerServiceClient;
+import com.google.cloud.secretmanager.v1beta1.SecretManagerServiceClient.ListSecretsPagedResponse;
+import com.google.cloud.secretmanager.v1beta1.SecretVersionName;
+
+import org.springframework.cloud.gcp.core.GcpProjectIdProvider;
+import org.springframework.core.env.EnumerablePropertySource;
+
+public class SecretManagerPropertySource extends EnumerablePropertySource<SecretManagerServiceClient> {
+
+	private static final String SECRETS_NAMESPACE = "spring-cloud-gcp.secrets.";
+
+	private static final String LATEST_VERSION_STRING = "latest";
+
+	private final Map<String, Object> properties;
+
+	private final String[] propertyNames;
+
+	public SecretManagerPropertySource(
+			String propertySourceName,
+			SecretManagerServiceClient client,
+			GcpProjectIdProvider projectIdProvider) {
+
+		super(propertySourceName, client);
+
+		Map<String, Object> propertiesMap = initializePropertiesMap(client, projectIdProvider.getProjectId());
+
+		this.properties = propertiesMap;
+		this.propertyNames = propertiesMap.keySet().toArray(new String[propertiesMap.size()]);
+	}
+
+	@Override
+	public String[] getPropertyNames() {
+		return propertyNames;
+	}
+
+	@Override
+	public Object getProperty(String name) {
+		return properties.get(name);
+	}
+
+	private static Map<String, Object> initializePropertiesMap(SecretManagerServiceClient client, String projectId) {
+		ListSecretsPagedResponse response = client.listSecrets(ProjectName.of(projectId));
+
+		HashMap<String, Object> secretsMap = new HashMap<>();
+		for (Secret secret : response.iterateAll()) {
+			String secretId = extractSecretId(secret);
+			if (secretId != null) {
+				String secretName = SECRETS_NAMESPACE + secretId;
+				String secretPayload = getSecretPayload(client, projectId, secretId);
+				secretsMap.put(secretName, secretPayload);
+			}
+		}
+
+		return secretsMap;
+	}
+
+	private static String getSecretPayload(
+			SecretManagerServiceClient client, String projectId, String secretId) {
+
+		SecretVersionName secretVersionName = SecretVersionName.newBuilder()
+				.setProject(projectId)
+				.setSecret(secretId)
+				.setSecretVersion(LATEST_VERSION_STRING)
+				.build();
+
+		AccessSecretVersionResponse response = client.accessSecretVersion(secretVersionName);
+		return response.getPayload().getData().toStringUtf8();
```

Comment:
Should probably document it in SpringBoot that only "string" secrets are accepted. Secret Manager accepts a wide character set for payload values.

Reply:
`public Object getProperty(String name)`
 So, I don't think we really need to convert all values to Strings. @dzou WDYT?

Reply:
Ah good point. I could just leave it as the [SecretPayload](https://cloud.google.com/secret-manager/docs/reference/rest/v1beta1/SecretPayload) which would be more flexible. Users would then be injecting/reading `SecretPayload` from the environment properties (instead of us forcing strings) and can convert the values to string or their preferred payload.

Do you feel this is a reasonable approach?

Reply:
Well, I think there might be problems with mapping to `@ConfigurationProperties` classes that have `Strings` and other basic Java types. It's something that we should test.

Reply:
You wouldn't want `SecretPayload` leaking into user code.

Reply:
Spring does the conversion automatically. If there isn't a `SecretPayload` encode/decoder it would fail.

Reply:
Ohh ok, this `SecretPayload` object can really just be converted to two options: `String` or `byte[]`.

@spencergibb - How do I hook into this encode/decoder concept? Like is it possible for me to override `getProperty(String propertyName, Class<T> targetType)` such that if targetType is `String` I decode to string, and if the target type is `byte[]` I decode to byte[]?

Reply:
String and byte[] properties are now supported through registering custom SecretPayload converters. See: https://github.com/spring-cloud/spring-cloud-gcp/pull/2168/files#diff-04e47cc423dee80f0eb313f527e9f5b2R74-R90